Implement Gillespie in C and call from python

prismofeverything opened the pull request:
This PR dispenses struggling with trying to make python fast and appeals directly to C, the source of all things. 

The python interface is the same (though it returns a dictionary instead of a tuple as the values returned were growing), but instead of using the now reference python implementation the arguments are translated from python into C arrays and passed to a pure C Gillespie implementation. Comparing the two:

```
reference time elapsed: 2.04260396957
obsidian time elapsed: 0.0289080142975
```

This is in the range of two orders of magnitude, or over 70x faster. Luckily for us, the Gillespie algorithm can never run too fast, so some day I hope to see this realized: https://www.ncbi.nlm.nih.gov/pubmed/19162916 (anyone know who the FPGA people are on campus? I have already asked several of you). Until then, we have C. Which will suffice for now.

I go into great detail about the design and algorithm throughout the code so I will avoid that here. The main entry point and translation from python occurs in `obsidianmodule.c` and the Gillespie implementation is in `obsidian.c` with a header at `obsidian.h`, in the `evolve` function. Any and all questions/suggestions/thoughts/revelations welcome, thanks!

jmason42 commented on arrow/obsidian.c:
```diff
@@ -0,0 +1,279 @@
+#include "obsidian.h"
+
+// Initial length of event vectors
+static int INITIAL_LENGTH = 4000;
```
If this becomes a performance bottleneck (which I highly doubt), I can probably come up with a heuristic for choosing a good size, based on the number of expected events for the simulated interval.

jmason42 commented on arrow/obsidian.c:
```diff
@@ -0,0 +1,279 @@
+
+  // Clean up the transient allocations
+  free(propensities);
+  free(update);
```
I know little about C and even less about connecting C with Python.  I'm sure it's a good practice but I'm curious about the necessity of these `free()` calls.  Would we leak memory without these?

prismofeverything replied:
To be fair, I was basing not freeing `time`, `events` and `outcome` on this documentation: https://docs.scipy.org/doc/numpy-1.15.0/reference/c-api.array.html#c.PyArray_NewFromDescr especially the big red "Warning" box that says not to deallocate pointers that are getting used to construct numpy arrays for passing back into python. Also it looks like the value NPY_OWNDATA is set which is still potentially unclear but to me reads that numpy takes over responsibility of calling `free` on these pointer. I may be wrong however and this should be verified. The other sign is that I did not observe the memory for the simulation growing as symptomatic of a memory leak though I have not done an official measurement on this. I'll try that soon and report back.

If I do need to call free on these pointers then I need to know when numpy is done with them, which seems difficult to achieve. Probably they would have to be wrapped in an object with an explicit deallocate method. Simpler perhaps now that I think of it would be to construct an empty numpy array from C and fill it with the values from those three arrays and then deallocate them. That will solve the problem if it is shown to be leaking, I'll investigate further.

1fish2 replied:
Ah. Memory management is subtle! So if numpy takes ownership, then it just needs a comment. I like the approach of constructing a numpy array.

jmason42 commented on arrow/obsidianmodule.c:
```diff
@@ -0,0 +1,468 @@
+#include <Python.h>
```
This file looks nightmarish.  Is there a good reference for developing this sort of Python-NumPy-C code?

1fish2 replied:
To @jmason42 's question about reference docs: also https://docs.python.org/2/c-api/index.html

prismofeverything commented on arrow/arrow.py:
```diff
@@ -1,146 +1,112 @@
+            self.dependencies_flat,
+            self.substrates_lengths,
+            self.substrates_indexes,
+            self.substrates_flat)
```
Yeah I thought about this but wanted to keep things simple implementation-wise, though a struct for the nested arrays could be warranted. The main thing that dissuaded me from this is how `reactants` and `reactions` shared the same `_lengths` and `_indexes` and it seemed honoring this would introduce more complexity than simply passing everything in. 

I can look into the cdef/cython thing, though at this point that sounds like a new implementation, and it is not guaranteed to not introduce other slowdowns we are not anticipating (my attempts with cython on this problem before showed minimal speedups, though this avenue was not explored exhaustively).

1fish2 replied:
Yeah the approach there is to let Cython do a thin glue layer and leave the work to C code. I know that's a big change for all the obsidianmodule.c code. The result ought to be a lot simpler and more maintainable...

prismofeverything replied:
Yeah that does sound simpler. I think since this is working I'm going to roll with it and if I ever need to come back here and modify anything instead of wrestling with the module code again I'll try out the cython-as-glue approach.

jmason42 commented on arrow/obsidian.c:
```diff
@@ -0,0 +1,279 @@
+evolve_result
+evolve(int reactions_count,
+       int substrates_count,
+       long * stoichiometry,
```
Incidentally I had to look up what the asterisk meant in this context, as I was only familiar with the `long *stoichiometry` form. Python has trained me to think that whitespace matters.

prismofeverything replied:
Ah you know this is an old habit but I tend to sometimes lose the `*` if it is jammed up against the word like that. I like to call out the pointer as loud as possible. Style contains the most stubborn arguments : ) I'll try it out and see if I can deal.

prismofeverything commented on MANIFEST.in:
```diff
@@ -0,0 +1 @@
+include arrow/obsidian.h
```
Well, it was `arrow` first and then we needed something faster, a kind of hard cutting mass at the tip of our arrow to do the real work, the business end. I have some dim memory that arrows used to be tipped with obsidian or something, or how it is still used today in certain specialist surgical implements. It was the first thing that came to me really, I tend to go with the flow on the naming thing and not spend too much time deliberating on it. Actually I think this is one of the more inspired : ) The practice of constantly having to name things is one of my favorite parts about programming.

1fish2 replied:
Perhaps add a comment, even if just to say it's a code name? `This invokes the Obsidian C module with the stoichiometry...` sounds like readers are expected to know what that is.

prismofeverything commented on arrow/obsidian.c:
```diff
@@ -0,0 +1,279 @@
+  long event_bounds = INITIAL_LENGTH;
+
+  // Allocate the dynamic arrays that will be used to track the progress of the system
+  double * time = malloc((sizeof (double *)) * event_bounds);
```
Ha, yeah! Bugs you don't notice because they still work, thanks @1fish2.

prismofeverything commented on arrow/obsidian.c:
```diff
@@ -0,0 +1,279 @@
+  // array we will use to actually apply the reactions and determine the next step's
+  // state.
+  for (species = 0; species < substrates_count; species++) {
+    outcome[species] = state[species];
```
I use `memcpy()` below and neglect it here, I remembered it halfway through it seems : )

prismofeverything commented on arrow/obsidian.c:
```diff
@@ -0,0 +1,279 @@
+
+// Print an array of doubles
+int
+print_array(double * array, int length) {
```
Yeah this was purely debug stuff. I'm keeping it around out of nostalgia and in case I ever need to modify this it is fairly useful (though there is probably some logging library I could use instead.... trying to keep it simple).

prismofeverything commented on arrow/arrow.py:
```diff
@@ -1,146 +1,112 @@
+            self.substrates_flat)
+
+    def evolve(self, duration, state):
+        steps, time, events, outcome = self.obsidian.evolve(duration, state)
```
The result of the C module function is a tuple. It gets constructed here: https://github.com/CovertLab/arrow/blob/obsidian/arrow/obsidianmodule.c#L199-L203 (`Py_BuildValue` returns a tuple, the letters in that string "iOOO" are the type codes of the four elements of the tuple).

1fish2 commented:
Yes, pass in a `np.RandomState`, or pass in a seed and call NumPy, or pass in a seed and then call its own random number generator, e.g. crib NumPy's Mersenne Twister. But finding, linking in, and testing a portable, high-quality generator sounds like a rabbit hole.

prismofeverything commented:
Okay, found a good BSD-licensed mersenne twister in C++ I was able to adapt to our purposes: https://github.com/cslarsen/mersenne-twister

The random state was static before so I added it to the obsidian object during initialization and added it as an argument to all the mersenne functions. Now each new obsidian object will have its own random state that it can track and keep independent of other states throughout the simulation. I also added a few functions, `sample_uniform` and `sample_expontential` to convert the unsigned ints into the doubles we need: https://github.com/CovertLab/arrow/commit/90dc3f58dfed2ba96bf79e5c0cd20f023d5bd93d

The interface has acquired a new argument to pass in the seed. 

I ran it a few times and the results are entirely identical. I haven't tested it with the sim yet but I assume this holds true.

prismofeverything commented on arrow/arrow.py:
```diff
@@ -1,146 +1,112 @@
+from .reference import derive_reactants, calculate_dependencies
+import obsidian
+
+def flatten(l):
```
> A shorter, more functional approach is `sum(l, [])`.

Nice, hadn't seen that one. Thanks!

prismofeverything commented on arrow/arrow.py:
```diff
@@ -1,146 +1,112 @@
+    flat = np.array(flatten(variable))
+    return flat, lengths, indexes
+
+def reenact_events(stoichiometry, events, state):
```
`events` is just a list of each reaction that occurred, as an index into the stoichiometry. You can use this to replay events on the initial state, reproducing the history of the system. I'm explaining this more in the next commit.

prismofeverything commented on arrow/obsidian.c:
```diff
@@ -0,0 +1,279 @@
+      // Find which reaction to update and initialize the propensity for that reaction
+      // with the rate for that reaction
+      reaction = update[up];
+      propensities[reaction] = rates[reaction];
```
Also, in all the cases `reaction` is used it refers to the *same index*, so seems apropos to me.

1fish2 replied:
Well, I totally missed that assignment to `reaction` on the previous line!

BTW I'm not so picky about `while (...)` and `sizeof(...)` as with the `*`s. As you noted, pointers are very significant and easy to goof. (At least we don't have the nasty case where it's not obvious what the code is incrementing when it has a pointer to an array of image bitplanes, which are arrays of scanlines, which are arrays of pixels.)

FYI, it's common to write a loop increment like `p++` but I don't know why. The postfix increment operator is a strange beast "return the variable's contents and by the way side-effect increment the variable." Personally I think it's better practice to use it only when needed, like `*p++` (which is a PDP-11 addressing mode), otherwise use `++p` which straightforwardly says "increment this variable". But it's OK either way.

prismofeverything commented:
Okay, handling malloc errors now, what do you think @1fish2? https://github.com/CovertLab/arrow/commit/e0017e8ecef0d390ae07a7c8f39c6bf7e6a5d9a7

I'm not sure if I need to free the other allocated items if it fails since that means the program is likely ending soon anyway, but thought I would be careful there. The way to signify a python call has failed is to return `NULL` from it, so this will escalate to an error in the module code if malloc fails here, which I think is the correct behavior. Open to feedback here of course.
